kolena/annotation.py

```diff
@@ -382,6 +387,123 @@ class ScoredLabeledTimeSegment(TimeSegment):
     """The score associated with this time segment."""
 
 
+@dataclass(frozen=True, config=ValidatorConfig)
+class TextSegment(DataObject):
+    """
+    Segments of text in the associated text field
+    """
+
+    text_field: str
+    """Text field column name containing the text segments."""
+
+    segments: List[Tuple[int, int]]
+    """List of text segments. The first item of each tuple is the start index of the segment,
+    and the second item is the end index."""
+
+    def __post_init__(self) -> None:
+        if not self.segments:
+            raise ValueError("At least one text segment must be provided")
+        for start, end in self.segments:
+            if start < 0:
+                raise ValueError(f"Start index must be non-negative ({start} provided)")
+            if end < 0:
+                raise ValueError(f"End index must be non-negative ({end} provided)")
+            if start > end:
+                raise ValueError(f"Start index must be less than or equal to end index ({start} > {end})")
+
+
+def find_substring_indices(field_name: str, text: str, substrings: list[str]) -> Optional[TextSegment]:
+    indices = []
+    for substring in substrings:
+        escaped_search = re.escape(substring)
+
+        pattern = rf"\b{escaped_search}\b"
+
+        matches = re.finditer(pattern, text, re.IGNORECASE)
+
+        indices.extend([(match.start(), match.end()) for match in matches])
+
+    if indices:
+        return TextSegment(field_name, indices)
+    return None
+
+
+@dataclass(frozen=True, config=ValidatorConfig)
+class LabeledTextSegments(Annotation):
```

Comment:
Should extend `TextSegments`

Reply:
I am actually following the example for [Curve](https://github.com/kolenaIO/kolena/blob/a72f0933e5c3a4c63db67dac1856bacc1ce7616b/kolena/workflow/plot.py#L75) and [CurvePlot](https://github.com/kolenaIO/kolena/blob/a72f0933e5c3a4c63db67dac1856bacc1ce7616b/kolena/workflow/plot.py#L110), Curve is the building block for CurvePlot, and Curve extended DataObject while CurvePlot extended Plot. TextSegments will never be used on its own I believe

Reply:
Do you think this is the right pattern?